pallene/coder.lua

```diff
@@ -392,10 +392,10 @@ function Coder:pallene_entry_point_declaration(f_id)
     table.insert(args, {"Udata *"     , "G",    ""})
     table.insert(args, {"StackValue *", "base", ""})
 
-    for i, upval in ipairs(func.captured_vars) do
-        local typ, c_name, comment = self:prepare_captured_var(i, upval.decl)
-        table.insert(args, {ctype(typ), c_name, comment})
+    if #func.captured_vars >= 1 then
+        table.insert(args, {"TValue*", "U", C.comment("upvalues")})
```

Comment:
Having the `U` argument be optional migth be a bit awkward when we implement indirect Pallene calls in #354. I think we'll have 3 options in that case:

1. The function doing the indirect Pallene call will have to include an if statement to see if it calls passing the U or not (urgh!)
2. We'll need to create an "calldyn pallene entry point", which always receives the U, and which lives between the lua entry point and the "callstatic pallene entry point"
3. We could have the pallene entry point always receive the U parameter.


For simplicity, I'm inclined to go with option (3) at first, and then open a github issue asking if it would be better to switch to option (2). Since option (2) is more complicated, I would only like to do it if we have benchmarks saying it's faster.

Reply:
I agree overall, but is it okay to pass `NULL` when we're doing a "CallStatic" for instance?

Reply:
That's another idea, let's call it option (4) :)

Reply:
But for this PR I think we should still go with option (3), which is the simplest. We can reconsider options 2,3,4 in a future PR.

Reply:
I went with (4) for now, should I revert it back to (3). (Is there a benefit to doing that?)

Reply:
I suppose that the only benefit of 3 instead of 4 is being marginally simpler, one less conditional.

From the point of view of code review, the only thing I have a stronger opinion about is that I don't like the default argument in `call_pallene_function`. For example, it's weird that the argument order of the `xs` and the `upvalue_list` is different than the order they appear in the pallene entry point.